envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/RBACFilterFactory.kt

```diff
@@ -52,95 +53,146 @@ class RBACFilterFactory(
-            if (incomingEndpoint.clients.isEmpty()) {
-                logger.warn("An incoming endpoint definition for $serviceName does not have any clients defined." +
+    private fun getIncomingEndpointPolicies(
+        serviceName: String,
+        incomingPermissions: Incoming,
+        snapshot: GlobalSnapshot,
+        roles: List<Role>
+    ): List<EndpointWithPolicy> {
+        val principalCache = mutableMapOf<ClientWithSelector, List<Principal>>()
+        return incomingPermissions.endpoints
+            .mapNotNull { incomingEndpoint ->
+                if (incomingEndpoint.clients.isEmpty()) {
+                    logger.warn("An incoming endpoint definition for $serviceName does not have any clients defined." +
                         "It means that no one will be able to contact that endpoint.")
-                return@forEach
+                    return@mapNotNull null
+                }
+
+                val clientsWithSelectors = resolveClientsWithSelectors(incomingEndpoint, roles)
+                val principals = clientsWithSelectors.flatMap { client ->
+                    principalCache.computeIfAbsent(client) { mapClientWithSelectorToPrincipals(it, snapshot) }
+                }.toSet()
+
+                if (principals.isNotEmpty()) {
+                    val policy = Policy.newBuilder().addAllPrincipals(principals)
+                    val combinedPermissions = rBACFilterPermissions.createCombinedPermissions(incomingEndpoint)
+                    policy.addPermissions(combinedPermissions)
+                    EndpointWithPolicy(incomingEndpoint, policy)
+                } else {
+                    null
+                }
             }
+    }
 
-            val clientsWithSelectors = resolveClientsWithSelectors(incomingEndpoint, incomingPermissions.roles)
-            val principals = clientsWithSelectors.flatMap { mapClientWithSelectorToPrincipals(it, snapshot) }.toSet()
+    data class Rules(val shadowRules: RBAC.Builder, val actualRules: RBAC.Builder)
 
-            if (principals.isNotEmpty()) {
-                val policyName = clientsWithSelectors.joinToString(",") { it.compositeName() }
-                val policy: Policy.Builder = clientToPolicyBuilder.computeIfAbsent(policyName) {
-                    Policy.newBuilder().addAllPrincipals(principals)
-                }
+    private fun getRules(
+        serviceName: String,
+        incomingPermissions: Incoming,
+        snapshot: GlobalSnapshot,
+        roles: List<Role>
+    ): Rules {
 
-                val combinedPermissions = createCombinedPermissions(incomingEndpoint)
-                policy.addPermissions(combinedPermissions)
-            }
-        }
+        val incomingEndpointsPolicies = getIncomingEndpointPolicies(
+                serviceName, incomingPermissions, snapshot, roles
+        )
+
+        val restrictedEndpointsPolicies = incomingEndpointsPolicies.asSequence()
```

Comment:
I have a feeling that I've asked that already but why do we need `asSequence()` everywhere?

Reply:
We have `asSequence()` only where it makes sense. Sequences are lazy iterables, just like Java streams. Lists are not lazy, they are fully eveluated before next step starts. 
So this code:
```kotlin
letters = listOf("a", "b", "c")
val result = listOf(1, 2, 3).map { it + 100 }.flatMap { number -> letters.flatMap { letter -> "$letter = $number" }  }.filter { it.beginsWith("b") } 
```
Will create every intermediate list:
* `val list1 = result.map { it + 100 }`
* `val list3 = letters.flatMap { letter -> "$letter = 101" }`
* `val list4 = letters.flatMap { letter -> "$letter = 102" }`
* `val list5 = letters.flatMap { letter -> "$letter = 103" }`
* `val list2 = list1.flatMap { number -> letters.flatMap { letter -> "$letter - $number" }  }`
* `val result = list2.filter { it.beginsWith("b") } `

but we are only interested in `result`, so its unnecessary

Wit sequences, everything is evaluating lazy, so we allocate only `result` list:
```
```kotlin
letters = listOf("a", "b", "c")
val result = listOf(1, 2, 3).asSequence().map { it + 100 }.flatMap { number -> letters.asSequence().flatMap { letter -> "$letter = $number" }  }.filter { it.beginsWith("b") }.toList()
```

So it's an optimisations.
You need to be careful using sequences though. You may achieve opposite results, degrading performance instead of improving it. For example doing something like this:
```kotlin
val someSequence = sequenceOf(1,2).map { costlyOperation(it) }

val someList = someSequence.toList()
val otherList = someSequence.forEach { println(it.something) }
```
you will call `costlyOperation` 4 times, but if you use lists, you will call it only 2 times.

General rule of thumb is to use sequences only with chain of operations, especially with many flatMaps, etc. It will improve performance when the sequences are big. And always store lists in variables, not sequences.

Reply:
oops, all this talk and I discovered that after some code changes 2 of 4 `asSequence()` usages doesn't actually improve anything (it also doesn't make it worse). I will delete it.

Reply:
I know how they work. I was wondering why we use `asSequence` there, doing another allocation, when we are filtering the whole list. It was misleading to me, I thought something is missing in the chain.

Reply:
Ok, I misunderstood you, my bad.